File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.py
from math import ceil


def solution(m, musicinfos):
    def convert(time):
        return int(time[:2]) * 60 + int(time[3:])
    
    answer = None
    sharp = {'C#': 'c', 'D#': 'd', 'F#': 'f', 'G#': 'g', 'A#': 'a'}
    
    for sh in sharp:
        m = m.replace(sh, sharp[sh])

    for info in musicinfos:
        start, end, title, melody = info.split(',')
        
        for sh in sharp:
            melody = melody.replace(sh, sharp[sh])
            
        start, end = convert(start), convert(end)
        length = end - start
        melody *= ceil(length / len(melody))
        
        if m not in melody[:length]:
            continue
            
        if answer is None or answer[1] < length or (answer[1] == length and answer[2] > start):
            answer = (title, length, start)
    
    return answer[0] if answer is not None else '(None)'

File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    
    public String solution(String m, String[] musicinfos) {
        String answerTitle = "(None)";
        int maxPlayTime = -1;
        int minStartTime = Integer.MAX_VALUE;

        Map<String, String> sharp = new HashMap<>();
        sharp.put("C#", "c");
        sharp.put("D#", "d");
        sharp.put("F#", "f");
        sharp.put("G#", "g");
        sharp.put("A#", "a");

        for (String key : sharp.keySet()) {
            m = m.replace(key, sharp.get(key));
        }

        for (String info : musicinfos) {
            String[] parts = info.split(",");
            String startStr = parts[0];
            String endStr = parts[1];
            String title = parts[2];
            String melody = parts[3];

            for (String key : sharp.keySet()) {
                melody = melody.replace(key, sharp.get(key));
            }

            int start = Integer.parseInt(startStr.substring(0, 2)) * 60 + Integer.parseInt(startStr.substring(3));
            int end = Integer.parseInt(endStr.substring(0, 2)) * 60 + Integer.parseInt(endStr.substring(3));
            int playTime = end - start;

            StringBuilder playedMelody = new StringBuilder();
            int melodyLength = melody.length();
            for (int i = 0; i < playTime; i++) {
                playedMelody.append(melody.charAt(i % melodyLength));
            }

            if (playedMelody.toString().contains(m)) {
                if (playTime > maxPlayTime) {
                    maxPlayTime = playTime;
                    minStartTime = start;
                    answerTitle = title;
                } else if (playTime == maxPlayTime) {
                    if (start < minStartTime) {
                        minStartTime = start;
                        answerTitle = title;
                    }
                }
            }
        }

        return answerTitle;
    }
}